src/batch/src/executor/join/lookup_join.rs

```diff
@@ -257,95 +266,147 @@ impl<P: 'static + ProbeSideSourceBuilder> LookupJoinExecutor<P> {
+
+            self.probe_side_source.reset();
+            for row_key in &row_keys {
+                self.probe_side_source.add_scan_range(row_key)?;
+            }
+
+            let probe_child = self.probe_side_source.build_source().await?;
+            let mut probe_side_stream = probe_child.execute();
+
+            let mut probe_side_chunk_exists = true;
+
+            // for each row in the build side chunk, has_match tells us if it has a match
+            // on the probe side table
+            let mut has_match = all_row_refs
+                .iter()
+                .map(|rows| vec![false; rows.len()])
+                .collect_vec();
+
+            // Keep looping until there are no more probe side chunks to get, then do 1 more loop
+            while probe_side_chunk_exists {
+                let probe_side_chunk = probe_side_stream.next().await;
+                probe_side_chunk_exists = probe_side_chunk.is_some();
+
+                let probe_side_chunk = if let Some(chunk) = probe_side_chunk {
+                    Some(chunk?.compact()?)
+                } else {
+                    None
+                };
+
+                for (i, row_refs) in all_row_refs.iter().enumerate() {
+                    // We filter the probe side chunk to only have the rows whose key datums
+                    // matches the key datums of the row_refs we're currently looking at
+                    let expr = self.create_expression(&row_keys[i], 0);
+                    let chunk = if let Some(chunk) = probe_side_chunk.as_ref() {
+                        let vis = expr.eval(chunk)?;
+                        let chunk = chunk.with_visibility(vis.as_bool().iter().collect());
+                        Some(chunk.compact()?)
+                    } else {
+                        None
+                    };
+
+                    // The handling of the different join types and which operation they should
+                    // perform under each condition is very complex. For each
+                    // row, the gist of it is:
+                    //
+                    // Inner Join: Join the cur_row and the probe chunk until the latter is None.
+                    //
+                    // Left Outer Join: Same as inner join but if the probe chunk is None and the
+                    //      row has no match yet, null-pad the current row and append it
+                    //
+                    // Left Semi Join: Check for matches until probe chunk is None or a match is
+                    //      found. If there were any matches by the end, append the row.
+                    //
+                    // Left Anti Join: Check for matches until probe chunk is None or a match is
+                    //      found. If there were no matches by the end, append the row.
+                    // TODO: Simplify the logic of handling the different join types
+                    for (j, cur_row) in row_refs.iter().enumerate() {
```

Comment:
Each group has same key. Do we still have to iterate over each row? If so, what's the benefit of grouping? 😇

Reply:
If grouping doesn't have improvements, we can simply

```rust
for row in build_chunk.rows() {
	let key = ...
	if key is new {
		self.probe_side_source.add_scan_range(key)?;
	}
}

while probe_side_chunk_exists {
	for row in build_chunk.rows() {
		...
	}
}
```

i.e., just dedup requests without grouping build-side chunk.

Reply:
And it is actually nested loop join? What about hash join or merge sort join? 🤔

Reply:
The main purpose of grouping is to minimize the number of RPC requests we make to scan the table. Rows with the same key datums will match with the same probe side rows, so we can group those rows together and make one RPC requests for them. Similarly, because multiple scan ranges can be on the same partition, we can group those scan ranges together and have the same worker scan them.

Reply:
Admittedly, this does add a lot of extra calculation time to evaluate the data chunks, which might end up being slower than ignoring partition grouping entirely. Unfortunately, it's difficult to check because Lookup Join is difficult to benchmarked...

Reply:
> Admittedly, grouping by partition requires adds a lot of calculation time because it has to evaluate the data chunks, which might make it slower overall.

Hmm, it may make it a little slower, but it's in memory computation, so compared with the rpc it's still much cheaper. I agree with @xxchan we can in fact improve it with hash join, but currently LGTM.